rclcpp_examples/src/topics/listener_single_msg.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright 2014 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <memory>
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/executors.hpp>
+#include <rclcpp/memory_strategies.hpp>
+#include <rclcpp/strategies/single_msg_memory_strategy.hpp>
+
+#include <std_interfaces/msg/int32.hpp>
+#include <std_interfaces/msg/string.hpp>
+
+using namespace rclcpp::single_msg_memory_strategy;
+using rclcpp::memory_strategies::static_memory_strategy::StaticMemoryStrategy;
+
+size_t messages_received = 0;
+
+void chatterCallback(const std_interfaces::msg::String::ConstSharedPtr & msg)
+{
+  printf("Address of received message: %p\n", msg.get());
+  ++messages_received;
+}
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+  rclcpp::memory_strategy::MemoryStrategy::SharedPtr memory_strategy = nullptr;
+  if (argc > 1) {
+    std::string argument(argv[1]);
+    if (argument == "static") {
+      printf("Setting memory allocation strategy to 'static'.\n");
+      memory_strategy = std::make_shared<StaticMemoryStrategy>(StaticMemoryStrategy());
+    } else if (argument == "dynamic") {
+      printf("Setting memory allocation strategy to 'dynamic'.\n");
+    } else {
+      fprintf(stderr, "Warning: unknown argument.\n");
+      printf("Setting memory allocation strategy to default (dynamic).\n");
+    }
+  }
+  if (!memory_strategy) {
+    memory_strategy = rclcpp::memory_strategy::create_default_strategy();
+  }
+
+  rclcpp::executors::SingleThreadedExecutor executor(memory_strategy);
+  auto node = rclcpp::Node::make_shared("listener_single_msg");
+
+  executor.add_node(node);
+
+  SingleMessageMemoryStrategy<std_interfaces::msg::String> msg_strategy;
+  auto msg_strategy_ptr =
+    std::make_shared<SingleMessageMemoryStrategy<std_interfaces::msg::String>>(
+    msg_strategy);
```

Comment:
These two statements could be simplified as:

```c++
auto msg_strategy_ptr = std::make_shared<SingleMessageMemoryStrategy<std_interfaces::msg::String>>();
```

or

```c++
auto msg_strategy_ptr = SingleMessageMemoryStrategy<std_interfaces::msg::String>::SharedPtr(
  new SingleMessageMemoryStrategy<std_interfaces::msg::String>());
```

The current pair of statements are sort of wasteful since it expands to essentially this:

```c++
SingleMessageMemoryStrategy<std_interfaces::msg::String> msg_strategy;
SingleMessageMemoryStrategy<std_interfaces::msg::String>::SharedPtr msg_strategy_ptr(new SingleMessageMemoryStrategy<std_interfaces::msg::String>());
*msg_strategy_ptr = msg_strategy;  // implicit copy
```